java/src/main/native/src/TableJni.cpp

```diff
@@ -22,12 +22,185 @@
+
+    JNIEnv* env = cudf::jni::get_jni_env(jvm);
+    size_t left_to_copy = size;
+    const char * copy_from = static_cast<const char*>(gpu_data);
+    while (left_to_copy > 0) {
+      long buffer_amount_available = current_buffer_len - current_buffer_written;
+      if (buffer_amount_available <= 0) {
+        // should never be < 0, but just to be safe
+        rotate_buffer(env);
+        buffer_amount_available = current_buffer_len - current_buffer_written;
+      }
+      long amount_to_copy = left_to_copy < buffer_amount_available ? left_to_copy : buffer_amount_available;
+      char * copy_to = current_buffer_data + current_buffer_written;
+
+      CUDA_TRY(cudaMemcpyAsync(copy_to, copy_from, amount_to_copy, cudaMemcpyDeviceToHost, stream));
+
+      copy_from = copy_from + amount_to_copy;
+      current_buffer_written += amount_to_copy;
+      total_written += amount_to_copy;
+      left_to_copy -= amount_to_copy;
+    }
+  }
+
+  void flush() override {
+    if (current_buffer_written > 0) {
+      if (is_stream_set) {
+        // only sync with the stream when we are going to use the data.
+        CUDA_TRY(cudaStreamSynchronize(stream));
+      }
+      JNIEnv* env = cudf::jni::get_jni_env(jvm);
+      handle_buffer(env, current_buffer, current_buffer_written);
+      current_buffer = nullptr;
+      current_buffer_len = 0;
+      current_buffer_data = nullptr;
+      current_buffer_written = 0;
+    }
+  }
+
+  size_t bytes_written() override {
+    return total_written;
+  }
+
+  void set_alloc_size(long size) {
+    this->alloc_size = size;
+  }
+
+private:
+  void rotate_buffer(JNIEnv* env) {
+    if (current_buffer != nullptr) {
+      if (is_stream_set) {
+        // only sync with the stream when we are going to use the data.
+        CUDA_TRY(cudaStreamSynchronize(stream));
+      }
+      handle_buffer(env, current_buffer, current_buffer_written);
+      env->DeleteGlobalRef(current_buffer);
+      current_buffer = nullptr;
+    }
+    jobject tmp_buffer = allocate_host_buffer(env, alloc_size, true);
+    current_buffer = env->NewGlobalRef(tmp_buffer);
+    current_buffer_len = get_host_buffer_length(env, current_buffer);
+    current_buffer_data = reinterpret_cast<char *>(get_host_buffer_address(env, current_buffer));
+    current_buffer_written = 0;
+  }
+
+  void handle_buffer(JNIEnv* env, jobject buffer, jlong len) {
+    env->CallVoidMethod(callback, handle_buffer_method, buffer, len);
+    if (env->ExceptionCheck()) {
+      throw std::runtime_error("handleBuffer threw an exception");
+    }
+  }
+
+  JavaVM* jvm;
+  jobject callback;
+  jmethodID handle_buffer_method;
+  jobject current_buffer = nullptr;
+  char * current_buffer_data = nullptr;
+  long current_buffer_len = 0;
+  long current_buffer_written = 0;
+  size_t total_written = 0;
+  long alloc_size = 10 * 1024 * 1024;
+  cudaStream_t stream = nullptr;
+  bool is_stream_set = false;
+};
+
+template <typename STATE>
+class jni_x_writer_handle final {
```

Comment:
Curious about the `x` in the name.  Would `jni_table_writer_handle` be more readable?